### PR TITLE
Update history to version 4.3.0

### PR DESCRIPTION
**Pull request by vslinko**
Continue of #10261

Totally rewritten typings based on https://github.com/mjackson/history/tree/v4.3.0

### CODE REVIEW

File: history/tsconfig.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
         "module": "commonjs",
         "target": "es6",
         "noImplicitAny": true,
-        "strictNullChecks": false,
+        "strictNullChecks": true,
```
Comment (vslinko) [inline history/tsconfig.json]: Is it ok?

File: history/index.d.ts [hunk 2]
```diff
@@ -1,140 +1,112 @@
...
-// types based on https://github.com/rackt/history/blob/master/docs/Terms.md
+export type Action = 'PUSH' | 'REPLACE' | 'POP';
+export type Basename = string;
+export type Pathname = string;
+export type Search = string;
+export type Hash = string;
+export type LocationState = any;
+export type Key = string;
 
-export as namespace History;
```
Comment (vslinko) [inline history/index.d.ts]: I have errors on this line:

```
history/index.d.ts(6,1): error TS1128: Declaration or statement expected.
history/index.d.ts(6,11): error TS1005: ';' expected.
history/index.d.ts(6,36): error TS1005: '{' expected.
```

So, I just removed  this line. Is it fine?

Comment (mhegazy) [inline history/index.d.ts]: are you using TS 2.0.3?

Comment (vslinko) [inline history/index.d.ts]: Ow, I installed dependencies from the master branch. Fixed.

**Comment by vslinko** [comment]
Hmm, React Router is broken because it depends on the old version. How to update history and keep RR work?

**Comment by aindlq** [comment]
see http://definitelytyped.org/guides/contributing.html#version

**Comment by yuit** [comment]
@vslinko Thank you for the contribution, could you fix the build break?

**Comment by vslinko** [comment]
To fix the build I should update typings of React Router from v2 to v4 but I have two issues:

- API of RRv4 is dramatically changed;
- I'm not using RR.

So, I can't update RR typings and fix the build.

**Comment by paulvanbrenk** [comment]
This PR seems abandoned, please re-open when you made the requested changes.

**Pull request closed without merging**